### PR TITLE
CMK Encryption support for resources in templates

### LINKED ISSUE #4145: CMK support for resources outside of core
As a complementary issue to #4142, we want to add CMK support for the remaining TRE resources, such as workspaces, workspace services, shared services, etc.

- make sure to update the CMK section in the docs accordingly.

### PR DESCRIPTION
**Pull request by yuvalyaron**
# Resolves #4145

## What is being addressed

Added support for CMK encryption for resources in templates.

## How is this addressed

- Workspace services use the CMK and encryption identity of the workspace to enable encryption for supporting resources.
- Shared services use the CMK and encryption identity of the core TRE to enable encryption for supporting resources.

### CODE REVIEW

**Comment by marrobi** [comment]
@yuvalyaron do we know what the upgrade behaviour is here? Might be worth incrementing bundles by a major version to prevent easy upgrade?

**Review by LizaShak — APPROVED** [review]
LGTM

**Review by LizaShak — COMMENTED** [review]
Let me know when ready :)

**Comment by yuvalyaron** [comment]
> @yuvalyaron do we know what the upgrade behaviour is here? Might be worth incrementing bundles by a major version to prevent easy upgrade?

Upgrading is not an issue as long as the `enable_cmk_encryption` variable remains unchanged, and we assume it will never change (though the exact approach to how to ensure this is yet to be determined):  https://github.com/microsoft/AzureTRE/issues/4172

File: templates/workspace_services/guacamole/user_resources/guacamole-azure-linuxvm/terraform/main.tf [hunk 1]
```diff
@@ -73,3 +73,15 @@ data "azurerm_public_ip" "app_gateway_ip" {
   name                = "pip-agw-${var.tre_id}"
   resource_group_name = data.azurerm_resource_group.core.name
 }
+
+data "azurerm_key_vault_key" "ws_encryption_key" {
```
Comment (LizaShak) [inline templates/workspace_services/guacamole/user_resources/guacamole-azure-linuxvm/terraform/main.tf:77]: Maybe worth adding it in a data.tf file like it is everywhere else?

Comment (yuvalyaron) [reply templates/workspace_services/guacamole/user_resources/guacamole-azure-linuxvm/terraform/main.tf]: Done - created a data.tf file and moved everything

**Review by tamirkamara — APPROVED** [review]

**Comment by yuvalyaron** [comment]
The workspace tests fail because the Terraform code from the base workspace in the old release does not include the newly added variables. This issue should be resolved once a new release is created.

**Pull request merged**